calico/felix/endpoint.py

```diff
@@ -36,26 +38,49 @@ class Address(object):
     An address as reported to Felix by the plugin (ignoring fields that Felix
     does not use). The input *fields* parameter is the fields from the API
     ENDPOINTCREATED (or other ENDPOINT*) request, of which we just need the
-    address.
+    address and gateway.
 
     Raises KeyError if there is a missing "addr" field in the parameters.
+
     """
     def __init__(self, fields):
-        #*********************************************************************#
-        #* This constructor throws InvalidAddress if there is a missing or   *#
-        #* invalid field.                                                    *#
-        #*********************************************************************#
+        # Two things can go wrong here, addr key doesn't exist, or the IP
+        # address is not valid.  In both cases, the caller will deal with the
+        # exception thrown.
         try:
-            self.ip = fields['addr'].encode('ascii')
+            self.ip = IPAddress(fields['addr'])
         except KeyError:
-            raise InvalidAddress
+            raise InvalidAddress()
+
+        self.version = self.ip.version
 
-        if common.validate_ip_addr(self.ip, 4):
-            self.type = futils.IPV4
-        elif common.validate_ip_addr(self.ip, 6):
-            self.type = futils.IPV6
+        # The gateway is used by Felix to set up NDP proxy.  It is required
+        # for IPv6 addresses unless the deployment has some other mechanism
+        # for endpoints to learn the gateway's MAC address, such as Router
+        # Advertisements from a DHCP6 service (typical OpenStack mechanism).
+        # However, we would always expect the gateway to be provided on the
+        # API. lIf no gateway is provided, log a warning.
+        try:
+            self.gateway = IPAddress(fields['gateway'])
+        except KeyError:
+            log.warning("No gateway provided with address")
+            self.gateway = None
         else:
-            raise InvalidAddress
+            # If, for some reason, the IP version on the address and gateway
+            # don't match, we will fail to program the proxy.  So, better to
+            # fail the API request now.
+            if self.gateway.version != self.ip.version:
+                raise InconsistentIPVersion(
+                    "IP %s is not the same IP version as gateway %s" %
+                    (self.ip, self.gateway))
+
+
+class InconsistentIPVersion(Exception):
+    """
+    Tried to create an Address with inconsistent IP versions between
+    properties, e.g. IP and gateway.
+    """
+    pass
```

Comment:
Oof, nevermind, no such module exists. I'd like to have one with `InvalidAddress` in it, but if you are opposed to the idea I don't consider it blocking for this PR.